pyro/infer/abstract_infer.py

```diff
@@ -70,4 +80,43 @@ def run(self, *args, **kwargs):
         for tr, logit in poutine.block(self._traces)(*args, **kwargs):
             self.exec_traces.append(tr)
             self.log_weights.append(logit)
+        self._categorical = Categorical(logits=torch.tensor(self.log_weights))
         return self
+
+
+class TracePredictive(TracePosterior):
+    """
+    Generates and holds traces from the posterior predictive distribution,
+    given model execution traces from the approximate posterior. This is
+    achieved by constraining latent sites to randomly sampled parameter
+    values from the model execution traces and running the model forward
+    to generate traces with new response ("_RETURN") sites.
+
+    :param model: arbitrary Python callable containing Pyro primitives.
+    :param TracePosterior posterior: trace posterior instance holding
+        samples from the model's approximate posterior.
+    :param int num_samples: number of samples to generate.
+    :param list hide_nodes: list of nodes that should be hidden when
+        replaying the model against ``model_traces``. Corresponds to
+        observed nodes when sampling from the posterior predictive.
+    """
+    def __init__(self, model, posterior, num_samples, hide_nodes=[]):
+        self.model = model
+        self.posterior = posterior
+        self.hide_nodes = hide_nodes
+        self.num_samples = num_samples
+        super(TracePredictive, self).__init__()
+
+    def _get_random_pruned_trace(self):
+        trace = self.posterior()
+        for node in self.hide_nodes:
+            trace.remove_node(node)
```

Comment:
I think you can just remove this method.  You can prune observed sites in `TracePosterior.__call__` and include observed sites in `TracePredictive` by adding observations to `self.model` outside of `TracePredictive`.

Reply:
Yup, makes sense. Let me modify this and get the examples working.

Reply:
Removed obs sites by default. This looks much nicer.